components/support/sql/src/open_database.rs

```diff
@@ -0,0 +1,432 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// Use this module to open a new SQLite database connection.
+///
+/// Usage:
+///    - Define a struct that implements MigrationLogic.  This handles:
+///      - Initializing the schema for a new database
+///      - Upgrading the schema for an existing database
+///      - Extra preparation/finishing steps, for example setting up SQLite functions
+///
+///    - Call open_database() in your database constructor:
+///      - The first method called is setup_pragmas().  This is executed outside of a transaction
+///        to support `PRAGMA journal_mode=wal`
+///      - If the database file is not present, open_database() will create a new DB and call prepare(),
+///        init(), then finish()
+///      - If the database file exists, open_database() will open it and call prepare(),
+///        upgrade_from() for each upgrade that needs to be applied, then finish().
+///
+///  See the autofill DB code for an example.
+///
+use crate::ConnExt;
+use rusqlite::{Connection, OpenFlags, NO_PARAMS};
+use std::path::Path;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    // Generic error meaning that something went wrong during the migration.  You can return this
+    // from your upgrade functions to signal that the database is beyond repair and can't be
+    // migrated.
+    #[error("MigrationError: {0}")]
+    MigrationError(String),
+    // Error with the migration logic struct, for example the number upgrade functions doesn't
+    // line up with start_version and end_version
+    #[error("MigrationLogicError: {0}")]
```

Comment:
This doesn't seem used either? In fact, it looks like almost none of them are. How would you feel about avoiding the new error enum, having the only "real" error be the rusqlite one, and `VersionTooNew` could even just panic - there's nothing a consumer (or us!) could do to handle this in practice, and "less is more" WRT errors I think. I'm a bit on the fence about this though, so if you feel strongly let me know.

Reply:
I decided to leave in the error struct for 2 reasons:
  - It makes it easy to distinguish errors while opening the database vs errors during normal operations.  Maybe this will help us with tracking down where the corruption errors are happening
  - It feels a bit scary to me to panic while in the middle of a transaction.  This doesn't really apply to `VersionTooNew`, but it does apply to the current panics in the upgrade_from() function when the version is too old.

I'm also on the fence though.  1) could probably be handled a different way and 2) might be more paranoia than a real issue.